server/src/main/java/org/elasticsearch/readiness/ReadinessService.java

```diff
@@ -0,0 +1,226 @@
+    private static final Logger logger = LogManager.getLogger(ReadinessService.class);
+
+    private final Environment environment;
+
+    private volatile boolean active = false;
+    private volatile ServerSocketChannel serverChannel;
+    private CountDownLatch listenerThreadLatch;
+    final AtomicReference<InetSocketAddress> boundSocket = new AtomicReference<>();
+    private final Collection<BoundAddressListener> boundAddressListeners = new CopyOnWriteArrayList<>();
+
+    public static final Setting<Integer> PORT = Setting.intSetting("readiness.port", -1, Setting.Property.NodeScope);
+
+    public ReadinessService(ClusterService clusterService, Environment environment) {
+        this.serverChannel = null;
+        this.environment = environment;
+        clusterService.addListener(this);
+    }
+
+    // package private for testing
+    boolean ready() {
+        return this.serverChannel != null;
+    }
+
+    /**
+     * Checks to see if the readiness service is enabled in the current environment
+     * @param environment
+     * @return
+     */
+    public static boolean enabled(Environment environment) {
+        return PORT.get(environment.settings()) != -1;
+    }
+
+    // package private for testing
+    ServerSocketChannel serverChannel() {
+        return serverChannel;
+    }
+
+    /**
+     * Returns the current bound address for the readiness service.
+     * If Elasticsearch was never ready, this method will return null.
+     * @return the bound address for the readiness service
+     */
+    public BoundTransportAddress boundAddress() {
+        InetSocketAddress boundAddress = boundSocket.get();
+        if (boundAddress == null) {
+            return null;
+        }
+        TransportAddress publishAddress = new TransportAddress(boundAddress);
+        return new BoundTransportAddress(new TransportAddress[] { publishAddress }, publishAddress);
+    }
+
+    ServerSocketChannel setupSocket() {
+        assert PORT.get(environment.settings()) >= 0;
+        InetAddress localhost = InetAddress.getLoopbackAddress();
+        int portNumber = PORT.get(environment.settings());
+
+        AccessController.doPrivileged((PrivilegedAction<Void>) () -> {
+            try {
+                serverChannel = ServerSocketChannel.open(StandardProtocolFamily.INET);
+                if (boundSocket.get() != null) {
+                    serverChannel.bind(boundSocket.get());
+                } else {
+                    InetSocketAddress socketAddress = new InetSocketAddress(localhost, portNumber);
+                    serverChannel.bind(socketAddress);
+
+                    boundSocket.set((InetSocketAddress) serverChannel.getLocalAddress());
+
+                    BoundTransportAddress boundAddress = boundAddress();
+                    for (BoundAddressListener listener : boundAddressListeners) {
+                        listener.addressBound(boundAddress);
```

Comment:
Unless there is a specific reason to do so, then we should probably reduce the scope of this doPriv block, in particular to remove the listener callbacks from it. Really `bind` is the only operation that should require permission.

Reply:
Good point, I'll extract the code out.

Reply:
Ah I remember why it's in there now, I only want to send the addressBound the first time the socket is bound. The socket can come off and on again, but the event shouldn't be sent again. The listeners are invoked only on first bind, under the else statement.

Reply:
I'll refactor the code to make this happen.